dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -104,108 +106,65 @@ DMFileReader::DMFileReader(
+        const auto & pack_stats = dmfile->getPackStats();
+        pack_offset.resize(pack_stats.size());
+        {
+            size_t offset = 0;
+            for (size_t i = 0; i < pack_stats.size(); ++i)
+            {
+                pack_offset[i] = offset;
+                offset += pack_stats[i].rows;
+            }
+        }
+    }
+    // Initialize read_block_infos
+    initReadBlockInfos();
 }
 
 bool DMFileReader::getSkippedRows(size_t & skip_rows)
 {
     skip_rows = 0;
-    const auto & pack_res = pack_filter.getPackResConst();
     const auto & pack_stats = dmfile->getPackStats();
-    for (; next_pack_id < pack_res.size() && !pack_res[next_pack_id].isUse(); ++next_pack_id)
-    {
+    const auto end_pack_id = read_block_infos.empty() ? pack_stats.size() : read_block_infos.front().start_pack_id;
+    for (; next_pack_id < end_pack_id; ++next_pack_id)
         skip_rows += pack_stats[next_pack_id].rows;
-        addSkippedRows(pack_stats[next_pack_id].rows);
-    }
-    next_row_offset += skip_rows;
-    // return false if it is the end of stream.
-    return next_pack_id < pack_res.size();
+    addSkippedRows(skip_rows);
+
+    return !read_block_infos.empty();
 }
 
 // Skip the block which should be returned by next read()
 size_t DMFileReader::skipNextBlock()
 {
-    // find the first pack which is used
-    if (size_t skip_rows; !getSkippedRows(skip_rows))
-    {
-        // no block left in the stream
-        return 0;
-    }
-
-    // move forward next_pack_id and next_row_offset
-    const auto [read_rows, rs_result] = getReadRows();
-    if (read_rows == 0)
+    if (size_t skip_rows = 0; !getSkippedRows(skip_rows))
         return 0;
 
+    const auto [start_pack_id, pack_count, rs_result, read_rows] = read_block_infos.front();
+    read_block_infos.pop_front();
+    next_pack_id = start_pack_id + pack_count;
     addSkippedRows(read_rows);
-    scan_context->late_materialization_skip_rows += read_rows;
     return read_rows;
 }
 
-// Get the number of rows to read in the next block
-// Move forward next_pack_id and next_row_offset
-std::pair<size_t, RSResult> DMFileReader::getReadRows()
-{
-    const auto & pack_res = pack_filter.getPackResConst();
-    const size_t start_pack_id = next_pack_id;
-    const size_t read_pack_limit = getReadPackLimit(start_pack_id);
-    const auto & pack_stats = dmfile->getPackStats();
-    size_t read_rows = 0;
-    auto last_pack_res = RSResult::All;
-    for (; next_pack_id < pack_res.size() && pack_res[next_pack_id].isUse() && read_rows < rows_threshold_per_read;
-         ++next_pack_id)
-    {
-        if (next_pack_id - start_pack_id >= read_pack_limit)
-            break;
-        last_pack_res = last_pack_res && pack_res[next_pack_id];
-        read_rows += pack_stats[next_pack_id].rows;
-    }
-
-    next_row_offset += read_rows;
-    if (read_tag == ReadTag::Query && last_pack_res.allMatch())
-        scan_context->rs_dmfile_read_with_all += next_pack_id - start_pack_id;
```

Comment:
So `rs_dmfile_read_with_all` is also useless after these changes?

Reply:
Add it back.